### PR TITLE
Provide n_frames attribute to multi-frame formats.

### PR DESCRIPTION
**Pull request by anntzer**
cf #1190, #1192.  Tests missing.

### CODE REVIEW

**Comment by hugovk** [comment]
The build failed:

```
ERROR: TestFileTiff.test_seek
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python-pillow/Pillow/Tests/test_file_tiff.py", line 260, in test_seek
    im.seek(-1)
  File "/home/travis/build/python-pillow/Pillow/PIL/TiffImagePlugin.py", line 665, in seek
    raise ValueError("invalid seek")
ValueError: invalid seek
```

https://travis-ci.org/python-pillow/Pillow/builds/58532302

**Comment by anntzer** [comment]
Indeed, I changed the API so that negative seeks on TIFF now raise ValueError instead of interpreting them as seeks to 0 -- in line with the other file formats (see e.g. ImImagePlugin).  Perhaps this needs a deprecation period?  Or can I just change the test?

**Comment by hugovk** [comment]
We've normally been very wary of changing the API. PIL is almost 20 years old and we don't want to needlessly break people's code.

File: PIL/FliImagePlugin.py [hunk 1]
```diff
@@ -109,11 +110,35 @@ def _palette(self, palette, shift):
...
+            try:
+                while True:
+                    self.seek(self.tell() + 1)
+            except EOFError:
+                self._n_frames = self.tell() + 1
+            self.seek(current)
+        return self._n_frames
+
     def seek(self, frame):
+        if frame == self.__frame:
```
Comment (hugovk) [inline PIL/FliImagePlugin.py]: One `seek` function.

File: PIL/FliImagePlugin.py [hunk 2]
```diff
@@ -109,11 +110,35 @@ def _palette(self, palette, shift):
...
+            return
+        if frame < self.__frame:
+            self._seek(0)
+        for f in range(self.__frame + 1, frame + 1):
+            self._seek(f)
+
+    def seek(self, frame):
```
Comment (hugovk) [inline PIL/FliImagePlugin.py]: A second `seek` function.

**Comment by hugovk** [comment]
Untested code:

PIL/DcxImagePlugin.py (50.0%): Missing lines 67
PIL/FliImagePlugin.py (38.5%): Missing lines 115-123,126-131,160
PIL/GifImagePlugin.py (42.1%): Missing lines 95-103,107,109
PIL/ImImagePlugin.py (50.0%): Missing lines 265
PIL/MicImagePlugin.py (50.0%): Missing lines 76
PIL/MpoImagePlugin.py (50.0%): Missing lines 67
PIL/PsdImagePlugin.py (50.0%): Missing lines 137
PIL/SpiderImagePlugin.py (40.0%): Missing lines 135,159,171
PIL/TiffImagePlugin.py (52.6%): Missing lines 652-660

(via https://travis-ci.org/python-pillow/Pillow/jobs/59688402#L4197)

**Comment by anntzer** [comment]
Feel free to adopt the PR and write the tests, I know they are missing.  Probably just involves copy pasting an access to the n_frames attribute and a lot of subtly different seek patterns.

**Comment by hugovk** [comment]
Which is the correct seek function?

**Comment by radarhere** [comment]
I have exercised the suggested freedom to adopt the PR and write the tests.

**Comment by anntzer** [comment]
Go for it.

**Pull request closed without merging**